WT-16199 Revamp WT_KEY_PROVIDER extension interface

ablekhman opened the pull request:
- Remove flexible member array from `WT_CRYPT_KEYS` struct
- Use existing `WT_ITEM` as a holder for the keys
- Memory ownership is on the WT side; the key provider extension does not need to allocate/free keys memory

ablekhman committed:
WT-16199 Revamp WT_KEY_PROVIDER extension interface
- Remove flexible member array from `WT_CRYPT_KEYS` struct
- Use existing `WT_ITEM` as a holder for the keys
- Memory ownership is on the WT side; the key provider extension does not need to allocate/free keys memory

jiechenbo commented on src/include/wiredtiger.h.in:
```diff
@@ -5744,10 +5731,8 @@ struct __wt_crypt_keys {
         uint64_t lsn; /* Returns checkpoint LSN upon success. */
         int error;    /* Returns error code on failure. */
     } r;
```
Previously we made this into a union because we wanted to save some disk space. Now we plan to use WT_ITEM to write to disk space instead, can we remove the union?

ablekhman replied:
I still like the union because these fields are mutually exclusive. So having the union nicely conveys this idea that only one of them is valid at given time.

jiechenbo replied:
I remember @ddanderson  much preferred separating into different fields. Maybe it's best to see what he thinks?

jiechenbo commented on examples/c/ex_key_provider.c:
```diff
@@ -75,7 +75,7 @@ my_load_key(WT_KEY_PROVIDER *kp, WT_SESSION *session, const WT_CRYPT_KEYS *keys)
-    memcpy((uint8_t *)encryption_data, keys->data, keys->size);
+    memcpy((uint8_t *)encryption_data, keys->keys.data, keys->keys.size);
     my_kp->encryption_data = encryption_data;
     my_kp->checkpoint_lsn = keys->r.lsn;
```
On load() we don't need to provide a checkpoint_lsn. Encryption layer doesn't need this information

ablekhman replied:
I thought it's easy for us to do this. Therefore, the extension might find it useful, if only for logging right now. However, if providing the corresponding LSN is difficult, then we don't need the struct at all and can simply use `WT_ITEM`.

jiechenbo commented on src/include/wiredtiger.h.in:
```diff
@@ -5744,10 +5731,8 @@ struct __wt_crypt_keys {
     } r;
-    size_t size;    /* Size of encrypted key data field. */
-    uint8_t data[]; /* Encrypted keys. */
+    WT_ITEM keys;     /* Encrypted keys. */
```
I wonder if keys is the best name here, maybe data instead?

ablekhman replied:
`WT_ITEM` has `data` already. So it looks quite ugly:

```c
... = keys->data.data;
```

That's why I picked `keys`.

jiechenbo replied:
Hmm I see, how about `crypt_data`?

ablekhman committed:
WT-16199 Revamp WT_KEY_PROVIDER extension interface
- Remove flexible member array from `WT_CRYPT_KEYS` struct
- Use existing `WT_ITEM` as a holder for the keys
- Memory ownership is on the WT side; the key provider extension does not need to allocate/free keys memory

ablekhman committed:
WT-16199 Revamp WT_KEY_PROVIDER extension interface
- Remove flexible member array from `WT_CRYPT_KEYS` struct
- Use existing `WT_ITEM` as a holder for the keys
- Memory ownership is on the WT side; the key provider extension does not need to allocate/free keys memory

jiechenbo reviewed (approved):
LGTM!
